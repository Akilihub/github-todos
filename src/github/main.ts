/**
 * This is the entry point for your Probot App.
 * @param {import('probot').Application} app - Probot's Application class.
 */
export const main = (app: any) => {
    app.log('Cheers, the app runs on a server!')
  
    app.on('push', async (context: any) => {
      const postIssue: string = context.issue({body: 'Thanks for opening a repository in this account.'})
      return context.github.issues.createComment(postIssue)
    })
  
  }
  