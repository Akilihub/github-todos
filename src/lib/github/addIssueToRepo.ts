/**
 *  API Function that adds an issue with comments to a repository 
 */

 export interface IssueWithComments {
    title: string,
    comments: string[]
 }

 export default async function addIssueToRepo(context: any): Promise<IssueWithComments[]> {
    const octokit = context.github;

    const fields = {
      owner: "author",
      repo: "repository",
      title: "title",
      body: "comments"
    };

    const result = await octokit.issues.create(fields); // ** ??? get comments
    return result.issues.map( issue => {
      return {
        title: issue.title,
        comments: issue.comments
      }
    })
 }