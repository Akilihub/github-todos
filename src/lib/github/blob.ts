import fetch from "node-fetch";

export interface Blob {
    content: string;
    name: string;
}

const getFileContent = async (arrList): Promise<Blob[]> => {
  const promises: Array<Promise<Blob>> = arrList.map( async (file) => {
    const res = await fetch(file.url);
    const textBlob = await res.text();
    return ({
      content: textBlob,
      name: file.name
    });
  });
  return Promise.all(promises);
};

export default getFileContent;
