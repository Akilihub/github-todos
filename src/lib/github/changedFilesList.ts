/**
 * Gets files that have changed during a commit
 */
import { Context } from "./types";
import { getBasicRepoProps } from "./util";

export interface ModifiedFile {
  name: string;
  url: string;
}

export default async function getChangedFiles(context: Context): Promise<ModifiedFile[]> {
  const repoPropsObj = getBasicRepoProps (context);
  const sha = context.payload.head_commit.id;
  const octokit = context.github;
  const fields = {
    owner: repoPropsObj.owner,
    repo: repoPropsObj.repo,
    sha
  };
  const result = await octokit.repos.getCommit( fields);
  return result.files.map(file => ({name: file.filename, url: file.blob_url}));
}