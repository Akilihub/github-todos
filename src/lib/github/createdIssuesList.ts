/**
 * extracts a list of open issues created by the app
 */

export interface openIssues {
    title: string; // issue title
    id: string; // issue id
};

export default async function autoCreatedIssues (context: any): Promise<openIssues> {
    const octokit = context.github;
    const opts = {
      filter: 'subscribed', 
      state: 'open'
    }
    const result = await octokit.issues.getAll(opts);
    return result.files.map(file => ( {name: file.title, url: file.number} ));
}