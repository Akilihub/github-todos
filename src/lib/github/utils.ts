/**
 * function that returns basic repository properties from payload.
 */
import { groupBy, flatten, prop } from "ramda";

import { Context, Issue, RepoProps, GHIssue } from "./types";
import { RepoIssue } from "../parser";
import { ModifiedFile } from "./getFilesOnInstall";


export function getBasicRepoProps (context: Context): RepoProps {
  const owner = context.payload.repository.owner.login;
  const repo = context.payload.repository.name;
  const fields = { owner, repo };
  return fields;
}

// returns issue diff. Diff pre-existing issues on github with new ones
// from source code
export function diffIssues (repo: Issue[], githubIssues: GHIssue[]): Issue[] {
  const repoTitleGroup = groupBy(obj => obj.title, repo);
  return Object.keys(repoTitleGroup).map((repoIssueTitle) => {

    const diffedBody = githubIssues.map(ghIssue => {
      if (repoIssueTitle === ghIssue.title) {
        const repoIssueArr = repoTitleGroup[repoIssueTitle][0].body.split(`\n`);
        return repoIssueArr.filter(iss => !ghIssue.comments.includes(iss));
      }
      return null;
    })
    .filter(body => body !== null );

    const body = diffedBody.length
      ? flatten(diffedBody).join(`\n`)
      : repoTitleGroup[repoIssueTitle][0].body;

    return {title: repoIssueTitle, body};
  });
}

export function mergeFileRepoIssues(repoIssues: RepoIssue[][]): RepoIssue[] {
  const groupedByIssueType =
      groupBy<RepoIssue>(prop("keyWord"), flatten< RepoIssue>(repoIssues));
  const issueGroupsList: RepoIssue[][] =
      Object.keys(groupedByIssueType).map(key => groupedByIssueType[key]);
  return flatten<RepoIssue>(issueGroupsList);
}

export function getModifiedFiles(octokit, array, owner, repo): Promise<ModifiedFile[]> {
  const modifiedFiles: Array<Promise<ModifiedFile>> = array.data.files.map(async(file) => {
  const name = file.filename;
  const content = await octokit.repos.getContent({owner, repo, path: name});
  return { name, htmlUrl: content.data.html_url, downloadUrl: content.data.download_url};
});
  return Promise.all(modifiedFiles);
}

export async function getfileNames(octokit, owner, repo, sha) {
 const objFiles = await octokit.gitdata.getTree({owner, repo, sha});
 if (objFiles) return;
 return objFiles.tree.map( async (obj) => {
  if (obj.type !== "tree") {
    return {path: obj.path, url: obj.url};
      } else {
    return await getfileNames(octokit, owner, repo, obj.sha);
    }
  });
}