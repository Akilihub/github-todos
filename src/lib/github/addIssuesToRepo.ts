/**
 * API function to add comments to existing bot created issues.
 */
import { find } from "ramda";
import { getBasicRepoProps } from "./utils";

import { Context, Issue } from "./types";
import { getAllRepoIssues } from "./repoIssuesList";

export default async function addIssuesToRepo (context: Context, newIssues: Issue[]): Promise<void> {
  const octokit = context.github;
  const { owner, repo } = getBasicRepoProps (context);
  const ghIssues = await getAllRepoIssues(context);
  newIssues.forEach(async ({title, body}) => {
    const currentGHIssue = find(ghIssue => ghIssue.title === title, ghIssues);
    const fields = { // TODO: maybe adde assignees
      owner,
      repo,
      body,
      title,
      labels: ["GH-TODO-BOT"],
      ...(currentGHIssue ? { number: currentGHIssue.number } : {})
    };
    return currentGHIssue
      ? octokit.issues.createComment(fields)
      : octokit.issues.create(fields);
  });

}
