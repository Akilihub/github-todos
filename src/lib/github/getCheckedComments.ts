import { getBasicRepoProps } from "./utils";

import { Context } from "./types";

export default async function getCheckedComments (context: Context): Promise<string[]> {
  const octokit = context.github;
  const { owner, repo } = getBasicRepoProps (context);
  const fields = {
      owner,
      repo,
      labels: ["GH-TODO-BOT"],
    };
  const arrayComments = await octokit.octokit.issues.getForRepo(fields);
  return arrayComments.map(obj =>
    obj.body.split(`\n`).filter( (comment: string) => comment.includes(`[x]`)));
}
