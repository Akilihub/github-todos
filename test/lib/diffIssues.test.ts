import "jest";
import { diffIssues } from "../../src/lib/github/utils";

const fromSource = [
    {
      "body": `-[ ]README.md: 14 alex testing 1 on the 2s
      -[ ]README.md: 14  see u`,
      "title": "FIXMEs from source code"
    },
    {
      "body": `-[ ]README.md: 12
  -[ ]README.md: 11 test
  -[ ]README.md: 40 check
  -[ ]README.md: 40 see`,
      "title": "TODOs from source code"
    }
  ];
const fromGithub = [
    {
      "body": `-[ ]README.md: 12
  -[ ]README.md: 11 test
  -[ ]README.md: 10 niggas in paris
  -[ ]README.md: 14 whistle`,
      "title": "TODOs from source code"
    },
    {
      "body": `-[ ]README.md: 14 alex testing 1 on the 2s
      -[ ]README.md: 15 look at u
      -[ ]README.md: 10 ok`,
      "title": "FIXMEs from source code"
    }
  ];

describe("all issues data tests", () => {
    it("should return blob data",  () => {
      const _data = diffIssues(fromSource, fromGithub);
      expect(_data).toMatchSnapshot();
    }, 10000);
});
