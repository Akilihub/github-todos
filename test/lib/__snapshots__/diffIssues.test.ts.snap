// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all issues data tests should return blob data 1`] = `
Array [
  Object {
    "body": "      -[ ]README.md: 14  see u",
    "title": "FIXMEs from source code",
  },
  Object {
    "body": "  -[ ]README.md: 40 check
  -[ ]README.md: 40 see",
    "title": "TODOs from source code",
  },
]
`;
